flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessor.java

```diff
@@ -0,0 +1,309 @@
+			// the preparations here are not placed in the constructor because all work in it
+			// must be executed after all operators are opened.
+			readingInputIndex = selectFirstReadingInputIndex();
+		}
+		if (readingInputIndex == -1) {
+			return InputStatus.NOTHING_AVAILABLE;
+		}
+
+		lastReadInputIndex = readingInputIndex;
+		InputStatus inputStatus = inputProcessors[readingInputIndex].processInput();
+		checkFinished(inputStatus, readingInputIndex);
+		return inputSelectionHandler.reportInputStatus(inputStatus, readingInputIndex);
+	}
+
+	private int selectFirstReadingInputIndex() {
+		// Note: the first call to nextSelection () on the operator must be made after this operator
+		// is opened to ensure that any changes about the input selection in its open()
+		// method take effect.
+		inputSelectionHandler.nextSelection();
+
+		isPrepared = true;
+
+		return selectNextReadingInputIndex();
+	}
+
+	private void checkFinished(InputStatus status, int inputIndex) throws Exception {
+		if (status == InputStatus.END_OF_INPUT) {
+			operatorChain.endHeadOperatorInput(getInputId(inputIndex));
+			inputSelectionHandler.nextSelection();
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
+		IOException ex = null;
+		for (InputProcessor<?> input : inputProcessors) {
+			try {
+				input.close();
+			} catch (IOException e) {
+				ex = ExceptionUtils.firstOrSuppressed(e, ex);
+			}
+		}
+
+		if (ex != null) {
+			throw ex;
+		}
+	}
+
+	private int selectNextReadingInputIndex() {
+		if (!inputSelectionHandler.isAnyInputAvailable()) {
+			fullCheckAndSetAvailable();
+		}
+
+		int readingInputIndex = inputSelectionHandler.selectNextInputIndex(lastReadInputIndex);
+		if (readingInputIndex == -1) {
+			return -1;
+		}
+
+		// to avoid starvation, if the input selection is ALL and availableInputsMask is not ALL,
+		// always try to check and set the availability of another input
+		if (inputSelectionHandler.shouldSetAvailableForAnotherInput()) {
+			fullCheckAndSetAvailable();
+		}
+
+		return readingInputIndex;
+	}
+
+	private void fullCheckAndSetAvailable() {
+		for (int i = 0; i < inputProcessors.length; i++) {
+			InputProcessor<?> inputProcessor = inputProcessors[i];
+			// TODO: isAvailable() can be a costly operation (checking volatile). If one of
+			// the input is constantly available and another is not, we will be checking this volatile
+			// once per every record. This might be optimized to only check once per processed NetworkBuffer
+			if (inputProcessor.networkInput.isApproximatelyAvailable() || inputProcessor.networkInput.isAvailable()) {
+				inputSelectionHandler.setAvailableInput(i);
+			}
+		}
+	}
+
+	private CompletableFuture<?> isAnyInputAvailable() {
+		if (inputSelectionHandler.isAnyInputAvailable()) {
+			return AVAILABLE;
+		}
+		final CompletableFuture<?> anyInputAvailable = new CompletableFuture<>();
+		for (int i = 0; i < inputProcessors.length; i++) {
+			if (!inputSelectionHandler.isInputFinished(i)) {
```

Comment:
The future will never complete if all inputs are finished, won't it?

Reply:
Would probably be better to use `CompletableFuture#any` in any case.

Reply:
> Would probably be better to use CompletableFuture#any in any case.

I was hoping to do just that, but we can not include all of the futures, so we would have to be constructing new array every time we access `CompletableFuture#anyOf(...)` - which might be quite often.